src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java

```diff
@@ -32,23 +35,64 @@ public RdnsLookupRequest(
   public void run() {
     this.lookupStartedMillis = System.currentTimeMillis();
     try {
-      String result = InetAddress.getByName(this.address).getHostName();
+      final List<String> verified_results = new ArrayList<>();
+      final List<String> unverified_results = new ArrayList<>();
+      final String[] bytes = this.address.split("\\.");
+      if (bytes.length == 4) {
+        final String reverseDnsDomain =
+                bytes[3]
+                        + "." + bytes[2] + "." + bytes[1] + "." + bytes[0]
+                        + ".in-addr.arpa";
+        String[] reverseDnsRecords =
+            DomainNameSystem.getRecords(reverseDnsDomain, "PTR");
+        for (String reverseDnsRecord : reverseDnsRecords) {
+          boolean verified = false;
+          String[] forwardDnsRecords =
+              DomainNameSystem.getRecords(reverseDnsRecord, "A");
+          for (String forwardDnsRecord : forwardDnsRecords) {
+            if (forwardDnsRecord.equals(this.address)) {
+              verified = true;
+              break;
+            }
+          }
+          if (verified) {
+            verified_results.add(reverseDnsRecord);
+          } else {
+            unverified_results.add(reverseDnsRecord);
+          }
+        }
+      }
       synchronized (this) {
-        this.hostName = result;
+        this.verifiedHostNames = verified_results;
+        this.unverifiedHostNames = unverified_results;
       }
-    } catch (UnknownHostException e) {
-      /* We'll try again the next time. */
+    } catch (NamingException e) {
+      /* The Onionoo field is omitted for both lookup failure and absence of
+       * a host name. We'll try again the next time. */
     }
     this.lookupCompletedMillis = System.currentTimeMillis();
     this.parent.interrupt();
   }
 
   public synchronized String getHostName() {
-    return hostName;
+    if (verifiedHostNames == null ) {
```

Comment:
Please always refer to members using `this.`. It is redundant, but it's more explicit when reading the code, and it's consistent with the current code.